aoc2018/src/main/scala/day13.scala

```diff
@@ -1,8 +1,144 @@
-// IDEA: make list of Point that have + \ /
-// cards are tuple of (coords, dir, intersection) e.g. (Point(1,1), (1,0), (0,0,1)) --> moving right (x + 1, y + 0) each tick
-// / --> transforms (x+1, y-1)
-// \ --> transforms (x-1, y+1)
-// + --> transforms(x+0, y+0) or / \
-//
-// compute the whole lot. If any coord point of card is identical, then return crash coords
-// track cards only, + / \ are fixed and only operate if card has same coords
\ No newline at end of file
+import scala.io.*
+import math.*
+import scala.collection.mutable
+import scala.collection.mutable.Stack
+
+
+object day13 extends App:
+
+  private val day: String =
+    this.getClass.getName.drop(3).init
+
+  private val start1: Long =
+    System.currentTimeMillis
+
+  def rotateVector[A](n: Int, s: Vector[A]): Vector[A] =
+    if s.isEmpty then s
+    else
+      val nbound: Int = n % s.length // skipping the full rotation rounds
+      if nbound < 0 then rotateVector(nbound + s.length, s)
+      else s.drop(nbound) ++ s.take(nbound)
+
+  case class Point(x: Int, y: Int):
+    def add(p2: Point): Point = Point(x + p2.x, y + p2.y)
+  sealed trait Switch
+  case object Straight extends Switch
+  case object Back extends Switch
+  case object Forward extends Switch
+  case object Intersect extends Switch
+
+  case class Rail(switch: Switch, loc: Point)
+  case class Cart(loc: Point, dir: Point, intersect: Vector[Switch]):
+
+    def update: Cart =
+      Cart(loc.add(dir), dir, intersect)
+
+    def computeSwitch(turn: Switch): Cart =
+      turn match
+        case Straight => straight.update
+        case Back => backwardSwitch.update
+        case Forward => forwardSwitch.update
+        case Intersect => handleIntersect.update
+    def forwardSwitch: Cart =
+      Cart(loc, Point(dir.y, dir.x), intersect)
+
+    def backwardSwitch: Cart =
+      Cart(loc, Point(-dir.y, -dir.x), intersect)
```

Comment:
could be refactored into a `def -(that: Point): Point` on the `Point` case class itself